File: 2022/day11/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

public class Solution {

    static class Monkey {
        List<Long> items = new ArrayList<>();
        char operator;
        String operand;
        int divisor;
        int ifTrue;
        int ifFalse;
        long inspections = 0;

        Monkey copy() {
            Monkey m = new Monkey();
            m.items = new ArrayList<>(this.items);
            m.operator = this.operator;
            m.operand = this.operand;
            m.divisor = this.divisor;
            m.ifTrue = this.ifTrue;
            m.ifFalse = this.ifFalse;
            m.inspections = 0;
            return m;
        }
    }

    static List<Monkey> parseMonkeys(String text) {
        List<Monkey> monkeys = new ArrayList<>();
        String[] blocks = text.trim().split("\n\n");

        Pattern numberPattern = Pattern.compile("\\d+");
        Pattern opPattern = Pattern.compile("new = old ([+*]) (\\w+)");

        for (String block : blocks) {
            String[] lines = block.trim().split("\n");
            Monkey monkey = new Monkey();

            // Parse starting items
            Matcher numMatcher = numberPattern.matcher(lines[1]);
            while (numMatcher.find()) {
                monkey.items.add(Long.parseLong(numMatcher.group()));
            }

            // Parse operation
            Matcher opMatcher = opPattern.matcher(lines[2]);
            if (opMatcher.find()) {
                monkey.operator = opMatcher.group(1).charAt(0);
                monkey.operand = opMatcher.group(2);
            }

            // Parse divisor
            numMatcher = numberPattern.matcher(lines[3]);
            if (numMatcher.find()) {
                monkey.divisor = Integer.parseInt(numMatcher.group());
            }

            // Parse if true
            numMatcher = numberPattern.matcher(lines[4]);
            if (numMatcher.find()) {
                monkey.ifTrue = Integer.parseInt(numMatcher.group());
            }

            // Parse if false
            numMatcher = numberPattern.matcher(lines[5]);
            if (numMatcher.find()) {
                monkey.ifFalse = Integer.parseInt(numMatcher.group());
            }

            monkeys.add(monkey);
        }

        return monkeys;
    }

    static long applyOperation(long old, char operator, String operand) {
        long val = operand.equals("old") ? old : Long.parseLong(operand);
        if (operator == '+') {
            return old + val;
        } else {
            return old * val;
        }
    }

    static void simulate(List<Monkey> monkeys, int rounds, int reliefDivisor, boolean useModulo) {
        // Calculate product of all divisors for modulo
        long modValue = 1;
        if (useModulo) {
            for (Monkey m : monkeys) {
                modValue *= m.divisor;
            }
        }

        for (int round = 0; round < rounds; round++) {
            for (Monkey monkey : monkeys) {
                while (!monkey.items.isEmpty()) {
                    long item = monkey.items.remove(0);
                    monkey.inspections++;

                    // Apply operation
                    long newVal = applyOperation(item, monkey.operator, monkey.operand);

                    // Apply relief
                    if (reliefDivisor > 1) {
                        newVal /= reliefDivisor;
                    }

                    // Apply modulo to prevent overflow
                    if (useModulo) {
                        newVal %= modValue;
                    }

                    // Test and throw
                    if (newVal % monkey.divisor == 0) {
                        monkeys.get(monkey.ifTrue).items.add(newVal);
                    } else {
                        monkeys.get(monkey.ifFalse).items.add(newVal);
                    }
                }
            }
        }
    }

    static long monkeyBusiness(List<Monkey> monkeys) {
        List<Long> inspections = new ArrayList<>();
        for (Monkey m : monkeys) {
            inspections.add(m.inspections);
        }
        inspections.sort(Collections.reverseOrder());
        return inspections.get(0) * inspections.get(1);
    }

    static long part1(String text) {
        List<Monkey> monkeys = parseMonkeys(text);
        simulate(monkeys, 20, 3, false);
        return monkeyBusiness(monkeys);
    }

    static long part2(String text) {
        List<Monkey> monkeys = parseMonkeys(text);
        simulate(monkeys, 10000, 1, true);
        return monkeyBusiness(monkeys);
    }

    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get("../input.txt");

        String text = Files.readString(inputPath);

        System.out.println("Part 1: " + part1(text));
        System.out.println("Part 2: " + part2(text));
    }
}

File: 2022/day11/rust/solution.rs
use std::fs;

#[derive(Clone)]
struct Monkey {
    items: Vec<u64>,
    operator: char,
    operand: Option<u64>, // None means "old"
    divisor: u64,
    if_true: usize,
    if_false: usize,
    inspections: u64,
}

fn parse_monkeys(text: &str) -> Vec<Monkey> {
    let mut monkeys = Vec::new();

    for block in text.trim().split("\n\n") {
        let lines: Vec<&str> = block.lines().collect();

        // Parse items
        let items: Vec<u64> = lines[1]
            .split_once(": ")
            .unwrap()
            .1
            .split(", ")
            .map(|s| s.parse().unwrap())
            .collect();

        // Parse operation (e.g., "new = old * 19" or "new = old + old")
        let op_part = lines[2].split("new = old ").nth(1).unwrap();
        let mut parts = op_part.split_whitespace();
        let operator = parts.next().unwrap().chars().next().unwrap();
        let operand_str = parts.next().unwrap();
        let operand = if operand_str == "old" {
            None
        } else {
            Some(operand_str.parse().unwrap())
        };

        // Parse divisor
        let divisor: u64 = lines[3]
            .split_whitespace()
            .last()
            .unwrap()
            .parse()
            .unwrap();

        // Parse targets
        let if_true: usize = lines[4]
            .split_whitespace()
            .last()
            .unwrap()
            .parse()
            .unwrap();
        let if_false: usize = lines[5]
            .split_whitespace()
            .last()
            .unwrap()
            .parse()
            .unwrap();

        monkeys.push(Monkey {
            items,
            operator,
            operand,
            divisor,
            if_true,
            if_false,
            inspections: 0,
        });
    }

    monkeys
}

fn apply_operation(old: u64, operator: char, operand: Option<u64>) -> u64 {
    let val = operand.unwrap_or(old);
    match operator {
        '+' => old + val,
        '*' => old * val,
        _ => panic!("Unknown operator"),
    }
}

fn simulate(monkeys: &mut Vec<Monkey>, rounds: u32, relief_divisor: u64, use_modulo: bool) {
    // For part 2, use product of all divisors to keep numbers manageable
    let mod_value: u64 = if use_modulo {
        monkeys.iter().map(|m| m.divisor).product()
    } else {
        0
    };

    for _ in 0..rounds {
        for i in 0..monkeys.len() {
            while !monkeys[i].items.is_empty() {
                let item = monkeys[i].items.remove(0);
                monkeys[i].inspections += 1;

                // Apply operation
                let mut new_val = apply_operation(item, monkeys[i].operator, monkeys[i].operand);

                // Apply relief
                if relief_divisor > 1 {
                    new_val /= relief_divisor;
                }

                // Apply modulo to prevent overflow
                if use_modulo {
                    new_val %= mod_value;
                }

                // Test and throw
                let target = if new_val % monkeys[i].divisor == 0 {
                    monkeys[i].if_true
                } else {
                    monkeys[i].if_false
                };
                monkeys[target].items.push(new_val);
            }
        }
    }
}

fn monkey_business(monkeys: &[Monkey]) -> u64 {
    let mut inspections: Vec<u64> = monkeys.iter().map(|m| m.inspections).collect();
    inspections.sort_by(|a, b| b.cmp(a));
    inspections[0] * inspections[1]
}

fn part1(text: &str) -> u64 {
    let mut monkeys = parse_monkeys(text);
    simulate(&mut monkeys, 20, 3, false);
    monkey_business(&monkeys)
}

fn part2(text: &str) -> u64 {
    let mut monkeys = parse_monkeys(text);
    simulate(&mut monkeys, 10000, 1, true);
    monkey_business(&monkeys)
}

fn main() {
    let text = fs::read_to_string("../input.txt").expect("Could not read input file");

    println!("Part 1: {}", part1(&text));
    println!("Part 2: {}", part2(&text));
}
